Job failure policy controller support

clamoriniere1A opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/devel/pull-requests.md#the-pr-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/devel/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/devel/pull-requests.md#write-release-notes-if-needed
-->

**What this PR does / why we need it**:
Start implementing the support of the "Backoff policy and failed pod limit" in the ```JobController```  defined in https://github.com/kubernetes/community/pull/583.
This PR depends on a previous PR #48075  that updates the K8s API types.

TODO: 
* [X] Implement ```JobSpec.BackoffLimit``` support
* [x] Rebase when #48075 has been merged.
* [X] Implement end2end tests



implements https://github.com/kubernetes/community/pull/583

**Special notes for your reviewer**:

**Release note**:
```release-note
Add backoff policy and failed pod limit for a job
```

clamoriniere1A commented:
/assign @soltysh

dims commented:
/ok-to-test
/release-note-none

soltysh commented on pkg/controller/job/job_controller.go:
```diff
@@ -424,34 +457,37 @@ func (jm *JobController) syncJob(key string) error {
-			}(i)
+		jobFailed = true
+		newFailureConditionFunc = func(r record.EventRecorder) batch.JobCondition {
+			r.Event(&job, v1.EventTypeNormal, "DeadlineExceeded", "Job was active longer than specified deadline")
```
Should be `EventTypeWarning` not `EventTypeNormal`.

soltysh replied:
Don't create methods like this, just record event and save condition in tmp variable.

soltysh commented on pkg/controller/job/job_controller.go:
```diff
@@ -424,34 +457,37 @@ func (jm *JobController) syncJob(key string) error {
+	if jobHaveNewFailure && jobToManyFailed(previousRetry+1, job.Spec.BackoffLimit) {
+		jobFailed = true
+		newFailureConditionFunc = func(r record.EventRecorder) batch.JobCondition {
+			r.Event(&job, v1.EventTypeNormal, "BackoffLimitExceeded", "Job has reach the specified backoff limit")
```
Should be `EventTypeWarning` not `EventTypeNormal`.

soltysh replied:
Don't create methods like this, just record event and save condition in tmp variable.

clamoriniere1A replied:
I created 2 new string vars for "reason" and "message" used in by ```Event()```and ```NewCondition()```

soltysh commented on pkg/controller/job/job_controller_test.go:
```diff
@@ -80,6 +80,16 @@ func newJob(parallelism, completions int32) *batch.Job {
+	if backoffLimit >= 0 {
+		j.Spec.BackoffLimit = &backoffLimit
+	} else {
+		j.Spec.BackoffLimit = nil
```
This is generally not possible situation, since if everything goes through API it will get defaulted. So I'd just ignore this case.

clamoriniere1A replied:
block replaced by
```j.Spec.BackoffLimit = &backoffLimit```

soltysh commented on pkg/controller/job/job_controller_test.go:
```diff
@@ -80,6 +80,16 @@ func newJob(parallelism, completions int32) *batch.Job {
+	if backoffSeconds >= 0 {
+		j.Spec.BackoffSeconds = &backoffSeconds
+	} else {
+		j.Spec.BackoffSeconds = nil
```
Same here.

clamoriniere1A replied:
block replaced by
```j.Spec.BackoffSeconds = &backoffSeconds```

soltysh commented on pkg/controller/job/job_controller_test.go:
```diff
@@ -1217,3 +1247,54 @@ func bumpResourceVersion(obj metav1.Object) {
 	obj.SetResourceVersion(strconv.FormatInt(ver+1, 10))
 }
+
+func Test_jobToManyFailed(t *testing.T) {
```
Not needed, drop this.

clamoriniere1A replied:
yep

soltysh commented on pkg/controller/job/job_controller.go:
```diff
@@ -513,9 +548,46 @@ func (jm *JobController) syncJob(key string) error {
 	return manageJobErr
 }
 
+func jobToManyFailed(nbRetry int32, maxRetry *int32) bool {
```
See my previous comment, this method is not needed.

clamoriniere1A replied:
method removed

soltysh reviewed (changes requested):
I left you some comments.

clamoriniere1A commented on pkg/controller/job/dynamicbackoff.go:
```diff
@@ -0,0 +1,182 @@
+}
+
+// DynamicBackoff used to
+// TODO: Move DynamicBackoff in go-client/utils/controlflow package or merge it with
```
ok sure,  as discussed on slack. I will create two new methods in the current Backoff struct:
1) ```GetWithNbRetry()``` which returns backoff duration and number of retry.
2) ```NextWithInitDuration```which allow providing the default init backoff duration.

clamoriniere1A commented on pkg/controller/job/job_controller.go:
```diff
@@ -424,34 +457,37 @@ func (jm *JobController) syncJob(key string) error {
+			return newCondition(batch.JobFailed, "BackoffLimitExceeded", "Job has reach the specified backoff limit")
+		}
+	} else if pastActiveDeadline(&job) {
 		// TODO: below code should be replaced with pod termination resulting in
```
done

clamoriniere1A commented on pkg/controller/job/job_controller.go:
```diff
@@ -424,34 +457,37 @@ func (jm *JobController) syncJob(key string) error {
+	jobHaveNewFailure := failed > job.Status.Failed
+
+	// check if the number of failed jobs increaded since the last syncJob
+	if jobHaveNewFailure && jobToManyFailed(previousRetry+1, job.Spec.BackoffLimit) {
```
right

clamoriniere1A commented on pkg/controller/job/job_controller.go:
```diff
@@ -424,34 +457,37 @@ func (jm *JobController) syncJob(key string) error {
+
+	jobHaveNewFailure := failed > job.Status.Failed
+
+	// check if the number of failed jobs increaded since the last syncJob
```
right, thanks

clamoriniere1A commented:
Hi @soltysh, I updated the PR with all your comments addressed.

clamoriniere1A commented:
/test pull-kubernetes-e2e-gce-gpu

clamoriniere1A commented:
/test pull-kubernetes-e2e-kops-aws
/test pull-kubernetes-e2e-gce-etcd3

clamoriniere1A commented:
/test pull-kubernetes-e2e-kops-aws

soltysh commented on staging/src/k8s.io/client-go/util/flowcontrol/backoff.go:
```diff
@@ -67,15 +68,45 @@ func (p *Backoff) Get(id string) time.Duration {
 
+// GetWithNbRetry the current backoff Duration and the number of previous retry
+func (p *Backoff) GetWithNbRetry(id string) (time.Duration, int32) {
+	p.Lock()
```
With this new method the other `Get` should be:
```go
func (p *Backoff) Get(id string) time.Duration {
    delay, _ := p.GetWithRetryNumber(id)
    return delay
}
```
We don't want copy & paste here.

clamoriniere1A replied:
right, usually not me either.

soltysh commented on staging/src/k8s.io/client-go/util/flowcontrol/backoff.go:
```diff
@@ -67,15 +68,45 @@ func (p *Backoff) Get(id string) time.Duration {
+}
+
+// NextWithInitDuration move backoff to the next mark and increment the number of retry, capping at maxDuration
+func (p *Backoff) NextWithInitDuration(id string, backoffDuration time.Duration, eventTime time.Time) {
```
s/backoffDuration/initial

Let the name suggest what it's doing. Similarly to `NewBackOff`.

soltysh replied:
Again, do not copy&paste code. Create a helper function and call that from `Next` and `NextWithInitDuration` passing `p.defaultDuration` or `initial` accordingly.

soltysh commented on staging/src/k8s.io/client-go/util/flowcontrol/backoff.go:
```diff
@@ -67,15 +68,45 @@ func (p *Backoff) Get(id string) time.Duration {
 }
 
+// GetWithNbRetry the current backoff Duration and the number of previous retry
+func (p *Backoff) GetWithNbRetry(id string) (time.Duration, int32) {
```
`GetWithRetryNumber` seems better fitted name.

clamoriniere1A replied:
ok

soltysh commented on staging/src/k8s.io/client-go/util/flowcontrol/backoff.go:
```diff
@@ -27,6 +27,7 @@ import (
 type backoffEntry struct {
 	backoff    time.Duration
 	lastUpdate time.Time
+	retry      int32
```
@deads2k mind having a second look at these changes in BackOff?

soltysh commented on staging/src/k8s.io/client-go/util/flowcontrol/backoff.go:
```diff
@@ -136,9 +167,26 @@ func (p *Backoff) DeleteEntry(id string) {
 
+// StartBackoffGC used to start the Backoff garbage collection mechanism
+// Backoff.GC() will be executed each minute
+func StartBackoffGC(backoff *Backoff, stopCh <-chan struct{}) {
```
Not sure if this belongs here, but I'll defer to @deads2k for the final judgment.

soltysh reviewed (commented):
Controller changes lgtm, I've asked David to review backoff changes.

deads2k commented on staging/src/k8s.io/client-go/util/flowcontrol/backoff.go:
```diff
@@ -27,6 +27,7 @@ import (
 type backoffEntry struct {
 	backoff    time.Duration
 	lastUpdate time.Time
+	retry      int32
```
Can you pull this entire backoff.go change into a separate pull and get a review from the usual client suspects (chao, daniel, myself) with a good description of what its buying you. My initial impression is that is exposing internals that I wouldn't expect to see leaked.

clamoriniere1A replied:
@deads2k thanks for the review, I created the separated pull request: https://github.com/kubernetes/kubernetes/pull/51626

deads2k commented on pkg/controller/job/job_controller.go:
```diff
@@ -456,34 +492,28 @@ func (jm *JobController) syncJob(key string) error {
+	jobHaveNewFailure := failed > job.Status.Failed
+
+	// check if the number of failed jobs increased since the last syncJob
+	if jobHaveNewFailure && (previousRetry+1 > *job.Spec.BackoffLimit) {
```
Using it for this?  I'd keep track of my retries locally to start.  Of a few dozen uses, you've found one that want this feature and its not obvious to me that all backoffs are created equal.  

Also, does `AddRateLimited` on the workqueue not do what you want?

clamoriniere1A replied:
Looking at others controllers: ```federated.JobController``` and the ```Kubelet``` the usage of the ```flowcontrol.Backoff ``` was looking the best choice,

Adding a separated data-structure for storing the number of retries for each Job was a possibility that I envisaged. 
But the fact that we still need to update the ```Backoff``` structure in order to allow to set different initial delay durations for each entry, made me orient myself to integrate also the ```retry```counter in the ```Backoff``` struct

clamoriniere1A committed:
Job failure policy support in JobController

Job failure policy integration in JobController. From the
JobSpec.BackoffLimit the JobController will define the backoff
duration between Job retry.

It use the ```workqueue.RateLimitingInterface``` to store the number of
"retry" as "requeue" and the default Job backoff initial duration is set
during the initialization of the ```workqueue.RateLimiter.

Since the number of retry for each job is store in a local structure
"JobController.queue" if the JobController restarts the number of retries
will be lost and the backoff duration will be reset to 0.

Add e2e test for Job backoff failure policy

soltysh commented:
/lgtm
/approve

soltysh commented:
/approve no-issue

soltysh commented:
/lgtm

clamoriniere1A commented:
/test pull-kubernetes-e2e-kops-aws

soltysh commented:
/retest

soltysh commented:
Rebased so 

/lgtm
/approve

soltysh commented:
/retest

nikhiljindal commented:
The e2e test that was added in this PR ([sig-apps] Job should exceed backoffLimit) seems to be flaking recently: https://k8s-testgrid.appspot.com/release-1.8-blocking#gci-gce-1.8. 
@clamoriniere1A Can you please take a look?

clamoriniere1A commented:
@nikhiljindal yes sure
